runner.py

```diff
@@ -0,0 +1,200 @@
+
+
+async def get_remote_tags(
+    session: ah.ClientSession, image_ref: str, auth_token: Optional[str] = None
+) -> list[str]:
+    """Use the OCI Distribution API to get a list of tags for an image."""
+
+    user_url = _parse_image_ref(image_ref)
+    api_base = _get_api_base(image_ref)
+
+    target_url = "/".join([api_base, user_url.path, "tags/list"])
+
+    auth_token = auth_token or await docker_hub_auth(session, api_base)
+    headers = {
+        "Authorization": auth_token,
+        "Accept": "application/vnd.docker.distribution.manifest.v2+json",
+    }
+
+    async with session.get(target_url, headers=headers, raise_for_status=True) as response:
+        # FIXME: We should be expecting at least 404 & 403 here.
+        # TODO: This is actually a paginated API. Need `async yield` + multiple requests.
+        # Yeah, the typechecker has no way of confirming what's up.
+        # The remote API could change at any time. Such is life.
+        rv: list[str] = (await response.json()).tags
+        return rv
+
+
+async def get_remote_manifest(
+    session: ah.ClientSession, image_ref: str, tag: str, auth_token: Optional[str] = None
+) -> dict[str, Any]:
+    """Retrieve the manifest for a specific tag of an image."""
+
+    user_url = _parse_image_ref(image_ref)
+    api_base = _get_api_base(image_ref)
+
+    target_url = "/".join([api_base, user_url.path, "manifests", tag])
+
+    auth_token = auth_token or await docker_hub_auth(session, api_base)
+    headers = {
+        "Authorization": auth_token,
+        "Accept": "application/vnd.docker.distribution.manifest.v2+json",
+    }
+
+    async with session.get(target_url, headers=headers, raise_for_status=True) as response:
+        # FIXME: We should be expecting at least 403.
+        rv: dict[str, Any] = await response.json()
+        return rv
+
+
+def get_digest_blob_hash(manifest: dict[str, Any]) -> Optional[str]:
+    """Get the digest's hash from a manifest. The hash can then be used to get the digest blob."""
+    rv: Optional[str] = manifest.get("config", {}).get("digest", None)
+    return rv
+
+
+async def get_remote_blob(
+    session: ah.ClientSession, image_ref: str, blob_hash: str, auth_token: Optional[str] = None
+) -> dict[str, Any]:
+    """Given a blob's hash, retrieve the contents of that blob."""
+    user_url = _parse_image_ref(image_ref)
+    api_base = _get_api_base(image_ref)
+
+    target_url = "/".join([api_base, user_url.path, "blobs", blob_hash])
+
+    auth_token = auth_token or await docker_hub_auth(session, api_base)
+    headers = {
+        "Authorization": auth_token,
+        "Accept": "application/vnd.docker.distribution.manifest.v2+json",
+    }
+
+    async with session.get(target_url, headers=headers, raise_for_status=True) as response:
+        # FIXME: We should be expecting at least 403.
+        rv: dict[str, Any] = await response.json()
+        return rv
+
+
+def get_labels(digest_blob: dict[str, Any]) -> dict[str, str]:
```

Comment:
we should not make APIs that take or return `Any` (unless we actually mean it), its better to document the specific type we expect, and then later on the code that calls it may be able to return a better formed type than Any on its end

Reply:
So, this probably shouldn't accept `Any`, but `get_remote_blob` can actually return a couple different things, depending on the hash you give it. I've essentially split JSON Doc retrieval from parsing for both `get_remote_blob` and `get_remote_manifest`. And these are not simple JSON docs, unfortunately. I could spend a couple hours getting the full JSON schema encoded into python types, but frankly, I don't see how that's valuable to getting the bot done. Those calls are just going to get mocked out in unit tests anyways, and I'm not certain what sort of integration test we could write up for them that wouldn't break regularly as new images are published and labels change.